aws/resource_aws_codebuild_project.go

```diff
@@ -507,6 +557,24 @@ func flattenAwsCodebuildProjectArtifacts(artifacts *codebuild.ProjectArtifacts)
 	return &artifactSet
 }
 
+func flattenAwsCodebuildProjectCache(cache *codebuild.ProjectCache) []interface{} {
+	values := map[string]interface{}{}
+
+	if cache.Type != nil {
+		if *cache.Type == "NO_CACHE" {
```

Comment:
Have you considered making `NO_CACHE` the `Default` value of `type` field instead of "magically" turning it into empty string here?

I'm not sure where's the balance between hiding API's implementation details and helping the user to be honest - just food for thought.

Reply:
I wasn't aware of `Default` so I haven't even tried to go this route. But I do think it sounds a lot cleaner and transparent than what I came up with.

I gave it a go just now, however I couldn't make it work. I can't set `Default` for only the `type`, since `type` is required when `cache` is present. My next try was to give a `Default` to the whole `cache` block, but it won't allow me since it's a list. I figured in this case I should use `DefaultFunc` instead, however I can't make it work. I came up with 

```golang
func defaultAwsCodeBuildProjectCache() (interface{}, error) {
	values := make([]map[string]interface{}, 1)
	values[0]["type"] = "NO_CACHE"
	return values, nil
}
```
and I was getting `Error: aws_codebuild_project.foo: cache.0: expected object, got invalid` when trying to generate a plan containing a codebuild project with no cache defined. So I guessed maybe I need to allocate those objects since the might have been gc'ed or something like, so I tried:

```golang
func defaultAwsCodeBuildProjectCache() (interface{}, error) {
    defaultValue := make([]map[string]interface{}, 1)
    defaultValue[0] = make(map[string]interface{}, 1)
    defaultValue[0]["type"] = "NO_CACHE"
    return defaultValue, nil
}
```
But I'm still getting the same error as above. Any ideas on what might be wrong? Please bear in mind that this is my first real contact with Go whatsoever, so I might be stuck in something very basic :-)

Reply:
I gave it another try yesterday and came to the conclusion that it doesn't make sense to pursue this as a default, since "NO_CACHE" is not a user facing value that is going to be used in HCL. The behaviour of Codebuild projects in the AWS side is a bit weird in the sense that if you don't specify Cache you get a project without caching enabled, while once you set a caching option, the only way to remove it is by updating the project with "NO_CACHE" set. I don't think it makes sense to ask from the user that, once they enabled cached for their Codebuild project on Terraform, they must apply a configuration with "NO_CACHE" explicitly set in order for caching to be disabled. If I remove the cache block from my aws_codebuild_project I wouldn't expect for caching to still be enabled.

If anyone is waiting on me to merge, you can go ahead since I'm done with it unless anyone has some more feedback about it :-)